apache/cloudstack
# Feature: Forgot password

Docs PR: https://github.com/apache/cloudstack-documentation/pull/432
### Description

This PR adds the functionality to reset password for a user by email.

#### 8 Global settings:
Category | Type | Name | Default value | Description | Dynamic
-- | -- | -- | -- | -- | --
Advanced | Long | `password.reset.ttl` | 30 | Password reset ttl in minutes | Yes
Advanced | String | `password.reset.email.sender` | `null` | Password reset email sender | Yes
Advanced | String | `password.reset.smtp.host` | `null` | Password reset smtp host | No
Advanced | Integer | `password.reset.smtp.port` | 25 | Password reset smtp port | No
Advanced | Boolean | `password.reset.smtp.useAuth` | `False` | Use auth for smtp in Password reset | No
Advanced | String | `password.reset.smtp.username` | `null` | Password reset smtp username | No
Secure | String | `password.reset.smtp.password` | `null` | Password reset smtp password | No
Advanced | String | `password.reset.mail.template` | Hello {{username}}!\nYou have requested to reset your password. Please click the following link to reset your password:\n{{{reset_link}}}\nIf you did not request a password reset, please ignore this email.\n\nRegards,\nThe CloudStack Team | Password reset mail template. This uses mustache template engine. Available variables are: username, firstName, lastName, resetLink, token | Yes

#### 2 new APIs:
1. **Command**: _forgotPassword_
    **Params**: _username_, _domain_
    **Details**: Sends an email to the user with a token which can be used to reset the password using resetPassword command.
2. **Command**: resetPassword
    **Params**: _username_, _domain_, _token_, _password_
    **Details**: Resets the password for the user using the token generated via forgotPassword command.
















### Types of changes


### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale


### Screenshots (if appropriate):


### How Has This Been Tested?



1. Setup a mailserver or for testing purposes use the below command to setup maildev using docker
```bash
docker run -d -p 1080:1080 -p 1025:1025 maildev/maildev
```
2. Update the above global settings to point the smtp host to the above docker container and restart the management server.
3. Ensure an email is set for the test user
4. In UI, click on forgot password and enter details.
5. maildev runs a web ui to view all emails being sent. Open maildev's UI which is running on port 1080 and see if there are any emails.
6. There should be an email from cloudstack to reset the password for the user. Click on the link in the email to reset the password or copy the token.
7. Open the link in the email or navigate to `{ACS MS}/user/resetPassword` on the UI.
8. Enter the details along with token from the email. Set the new password.
9. Try login with the newly set password.


#### How did you try to break this feature and the system with this change?
